modules/api/php/endpoints/candidate/visit/dicoms.class.inc

```diff
@@ -138,20 +156,167 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+                'PSCID and CandID in POST data do not refer to the same candidate.'
+            );
+        }
+
+        $visit_label = $data['Visit'] ?? null;
+        if ($visit_label !== $this->_visit->getVisitLabel()) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'Visit label in URL and POST data do not match.'
+            );
+        }
+
+        $is_phantom = $data['IsPhantom'] ?? null;
+        if (!in_array($is_phantom, [true, false])) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'IsPhantom must be either `true` or `false`.'
+            );
+        }
+
+        $file_name    = $uploaded_file->getClientFilename();
+        $patient_name = $pscid . "_" . $input_candid . "_" . $visit_label;
+        if (!preg_match("/^{$patient_name}/", $file_name) && !$is_phantom) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                "Filename must start with $patient_name."
+            );
+        }
+
+        // *************************
+        // *** Upload file *********
+        // *************************
+        $incoming_path = \NDB_factory::singleton()
+            ->config()
+            ->getSetting('MRIUploadIncomingPath');
+
+        $incoming_dir = new \SplFileInfo($incoming_path);
+
+        $uploader = (new \LORIS\FilesUploadHandler($incoming_dir))
+            ->withPermittedMIMETypes(
+                'application/gzip',
+                'application/tar',
+                'application/tar+gzip',
+                'application/x-tar'
+            );
+
+        if ($request->getHeaderLine('LORIS-Overwrite') === 'overwrite') {
+            $uploader = $uploader->withOverwrite(true);
+        }
+
+        $upload_response = $uploader->handle($request);
+        if ($upload_response->getStatusCode() !== 201) {
+            return $upload_response;
+        }
+
+        // *************************
+        // *** Update Database *****
+        // *************************
+        $file_path = $incoming_dir->getPathname() . '/' . $file_name;
+        $file_info = new \SplFileInfo($file_path);
+        if (!$file_info->isFile()) {
+            return new \LORIS\Http\Response\JSON\InternalServerError(
+                'Upload failed (see error_log.)'
+            );
+        }
+
+        try {
+            $this->_visit->addMRIUpload(
+                $user,
+                $file_info,
+                $patient_name,
+                $is_phantom,
+            );
+        } catch (\Throwable $e) {
+            error_log($e->getMessage());
+            return new \LORIS\Http\Response\JSON\InternalServerError(
+                'Database insertion failed (see error_log.)'
+            );
+        }
+
+        // This concatenation is here because the API part is removed from the URL by
+        // the module, which may not be a good idea.
+        $path = '/api/v0.0.4-dev/' . $request->getUri()->getPath() . '/'
```

Comment:
mmm, not sure about hardcoding the API version here. @driusan @ridz1208?

In general, if users are using 0.0.3, then the endpoints would not be available and if they use 0.0.4-dev, then the endpoint would be available and the upload would work.

Reply:
This was like that in the original PR, I don't really know what I should do here TBH.